### PR TITLE
issue credential overhaul

### LINKED ISSUE #129: Implement alternative beginnings for credential exchange protocol
As a developer, I want to allow the Holder role to start a credential exchange with the credential proposal. See [Aries RFC 0036: Issue Credential Protocol 1.0](https://github.com/hyperledger/aries-rfcs/blob/master/features/0036-issue-credential/README.md)

### PR DESCRIPTION
**Pull request by TimoGlastra**
Signed-off-by: Timo Glastra <timo@animo.id>

- Removes the auto-complete from issue-credential for now as it was only a partial auto-complete and no option to opt-in or opt-out.
- Add credential proposal (Fixes #129)
- Also includes some updates to present proof for interoperabillity.
- Make test flows not depend on each other. In credentials.test.ts all tests dependend on the previous one. To my knowledge this is not best practice for a number of reasons (https://stackoverflow.com/a/55129587/10552895)

I'll try to make the PRs smaller after this one. But I think these changes were needed

### CODE REVIEW

File: src/lib/protocols/present-proof/messages/PresentationAckMessage.ts [hunk 1]
```diff
@@ -1,19 +1,16 @@
 import { Equals } from 'class-validator';
 
-import { AgentMessage } from '../../../agent/AgentMessage';
+import { AckMessage, AckMessageOptions } from '../../connections';
 import { PresentProofMessageType } from './PresentProofMessageType';
 
-interface PresentationAckMessageOptions {
-  id?: string;
-}
+export type PresentationAckMessageOptions = AckMessageOptions;
```
Comment (jakubkoci) [inline src/lib/protocols/present-proof/messages/PresentationAckMessage.ts:6]: The same question as with `CredentialAckMessageOptions`

**Review by jakubkoci — APPROVED** [review]
Just a few notes, but don't want to block :)

File: src/lib/__tests__/helpers.ts [hunk 2]
```diff
@@ -1,16 +1,21 @@
 /* eslint-disable no-console */
-// @ts-ignore
-import { poll } from 'await-poll';
 import logger from '../logger';
 import path from 'path';
 import { Subject } from 'rxjs';
 import { ConnectionRecord } from '../storage/ConnectionRecord';
-import { Agent, InboundTransporter, OutboundTransporter } from '..';
+import { Agent, InboundTransporter, OutboundTransporter, ProofEventType, ProofState, ProofStateChangedEvent } from '..';
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply src/lib/__tests__/helpers.ts:6]: They're the same? I'm not sure I understand

File: src/lib/protocols/issue-credential/messages/CredentialAckMessage.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { Equals } from 'class-validator';
+import { IssueCredentialMessageType } from './IssueCredentialMessageType';
+
+import { AckMessage, AckMessageOptions } from '../../connections';
+
+export type CredentialAckMessageOptions = AckMessageOptions;
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply src/lib/protocols/issue-credential/messages/CredentialAckMessage.ts:6]: Credential Ack message is basically an Ack message but with a different `@type`. So we reuse the `AckMessageOptions` but keep the freedom to change the types if it is going to divert from the base ack message in the future

File: src/lib/protocols/present-proof/messages/PresentationAckMessage.ts (hunk 1 above)
Comment (TimoGlastra) [reply src/lib/protocols/present-proof/messages/PresentationAckMessage.ts:6]: See answer above

**Pull request merged**